File: iOS/TUIKit/TUICallKit/TUICallKit/UI/Recents/TUICallRecordCallsCell.m
//
//  TUICallRecordCallsCell.m
//  TUICallKit
//
//  Created by noah on 2023/2/28.
//  Copyright © 2023 Tencent. All rights reserved.
//

#import "TUICallRecordCallsCell.h"
#import "Masonry.h"
#import "TUICallRecordCallsCellViewModel.h"
#import "TUICallingCommon.h"
#import "UIImageView+WebCache.h"
#import "TUIConfig.h"
#import "TUICallEngineHeader.h"
#import "TUIThemeManager.h"

static NSString *const gRecordCallsCellViewModelKVOAvatarImage = @"avatarImage";
static NSString *const gRecordCallsCellViewModelKVOFaceURL = @"faceURL";
static NSString *const gRecordCallsCellViewModelKVOTitle = @"titleLabelStr";

@interface TUICallRecordCallsCell ()

@property (nonatomic, assign) BOOL isViewReady;
@property (nonatomic, strong) UIImageView *avatarImageView;
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UIImageView *mediaTypeImageView;
@property (nonatomic, strong) UILabel *resultLabel;
@property (nonatomic, strong) UILabel *timeLabel;
@property (nonatomic, strong) UIButton *moreButton;
@property (nonatomic, strong) TUICallRecordCallsCellViewModel *viewModel;

@end

@implementation TUICallRecordCallsCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        _isViewReady = NO;
        self.contentView.backgroundColor = TUICallKitDynamicColor(@"callkit_recents_cell_bg_color", @"#FFFFFF");
    }
    return self;
}

- (void)didMoveToWindow {
    [super didMoveToWindow];
    if (_isViewReady) {
        return;
    }
    _isViewReady = YES;
    [self constructViewHierarchy];
    [self activateConstraints];
    [self bindInteraction];
}

- (void)dealloc {
    [self removeBindViewModel];
}

- (void)constructViewHierarchy {
    [self.contentView addSubview:self.avatarImageView];
    [self.contentView addSubview:self.titleLabel];
    [self.contentView addSubview:self.mediaTypeImageView];
    [self.contentView addSubview:self.resultLabel];
    [self.contentView addSubview:self.timeLabel];
    [self.contentView addSubview:self.moreButton];
}

- (void)activateConstraints {
    [self.avatarImageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self.contentView);
        make.leading.equalTo(self.contentView).offset(8);
        make.width.height.mas_equalTo(40);
    }];
    [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.contentView).offset(14);
        make.leading.equalTo(self.avatarImageView.mas_trailing).offset(8);
        make.trailing.lessThanOrEqualTo(self.timeLabel.mas_leading).offset(-20);
    }];
    [self.mediaTypeImageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.equalTo(self.contentView).offset(-14);
        make.leading.equalTo(self.avatarImageView.mas_trailing).offset(8);
        make.width.mas_equalTo(19);
        make.height.mas_equalTo(12);
    }];
    [self.resultLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self.mediaTypeImageView);
        make.leading.equalTo(self.mediaTypeImageView.mas_trailing).offset(4);
        make.width.mas_equalTo(100);
    }];
    [self.timeLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self.contentView);
        make.trailing.equalTo(self.moreButton.mas_leading).offset(-4);
        make.width.mas_equalTo(100);
    }];
    [self.moreButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self.contentView);
        make.trailing.mas_equalTo(-8);
        make.width.height.mas_equalTo(24);
    }];
}

- (void)bindInteraction {
    [self.moreButton addTarget:self action:@selector(moreButtonClick:) forControlEvents:UIControlEventTouchUpInside];
}

- (void)bindViewModel {
    [self.viewModel addObserver:self forKeyPath:gRecordCallsCellViewModelKVOAvatarImage options:NSKeyValueObservingOptionNew context:nil];
    [self.viewModel addObserver:self forKeyPath:gRecordCallsCellViewModelKVOFaceURL options:NSKeyValueObservingOptionNew context:nil];
    [self.viewModel addObserver:self forKeyPath:gRecordCallsCellViewModelKVOTitle options:NSKeyValueObservingOptionNew context:nil];
}

- (void)removeBindViewModel {
    @try {
        [self.viewModel removeObserver:self forKeyPath:gRecordCallsCellViewModelKVOAvatarImage context:nil];
        [self.viewModel removeObserver:self forKeyPath:gRecordCallsCellViewModelKVOFaceURL context:nil];
        [self.viewModel removeObserver:self forKeyPath:gRecordCallsCellViewModelKVOTitle context:nil];
    }
    @catch (NSException *exception) {
    }
}

#pragma mark - bind data

- (void)bindViewModel:(TUICallRecordCallsCellViewModel *)viewModel {
    self.viewModel = viewModel;
    [self bindViewModel];
    
    self.titleLabel.text = viewModel.titleLabelStr;
    self.resultLabel.text = viewModel.resultLabelStr;
    self.timeLabel.text = viewModel.timeLabelStr;
    [self.mediaTypeImageView setImage:[TUICallingCommon getBundleImageWithName:viewModel.mediaTypeImageStr]];
    [self.avatarImageView sd_setImageWithURL:[NSURL URLWithString:viewModel.faceURL] placeholderImage:viewModel.avatarImage];
    
    if (TUICallResultTypeMissed == viewModel.callRecord.result) {
        self.titleLabel.textColor = UIColor.redColor;
    } else {
        self.titleLabel.textColor = TUICallKitDynamicColor(@"callkit_recents_cell_title_color", @"#000000");
    }
}

#pragma mark - observeValueForKeyPath

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary<NSKeyValueChangeKey,id> *)change
                       context:(void *)context {
    TUICallRecordCallsCellViewModel *viewModel = (TUICallRecordCallsCellViewModel *)object;
    
    if ([keyPath isEqualToString:gRecordCallsCellViewModelKVOAvatarImage]) {
        [self.avatarImageView sd_setImageWithURL:nil placeholderImage:viewModel.avatarImage];
    } else if ([keyPath isEqualToString:gRecordCallsCellViewModelKVOFaceURL]) {
        [self.avatarImageView sd_setImageWithURL:[NSURL URLWithString:viewModel.faceURL] placeholderImage:nil];
    } else if ([keyPath isEqualToString:gRecordCallsCellViewModelKVOTitle]) {
        self.titleLabel.text = viewModel.titleLabelStr;
    }
}

#pragma mark - event action

- (void)moreButtonClick:(UIButton *)button {
    !self.moreBtnClickedHandler ?: self.moreBtnClickedHandler(self);
}

#pragma mark - getter and setter

- (UIImageView *)avatarImageView {
    if (!_avatarImageView) {
        _avatarImageView = [[UIImageView alloc] init];
        _avatarImageView.contentMode = UIViewContentModeScaleAspectFit;
        _avatarImageView.layer.cornerRadius = 20;
        _avatarImageView.clipsToBounds = YES;
    }
    return _avatarImageView;
}

- (UILabel *)titleLabel {
    if (!_titleLabel) {
        _titleLabel = [[UILabel alloc] initWithFrame:CGRectZero];
        _titleLabel.numberOfLines = 1;
        _titleLabel.font = [UIFont fontWithName:@"PingFangHK-Semibold" size:14];
        _titleLabel.textAlignment =  isRTL() ? NSTextAlignmentRight : NSTextAlignmentLeft;
    }
    return _titleLabel;
}

- (UIImageView *)mediaTypeImageView {
    if (!_mediaTypeImageView) {
        _mediaTypeImageView = [[UIImageView alloc] init];
        _mediaTypeImageView.contentMode = UIViewContentModeScaleAspectFit;
    }
    return _mediaTypeImageView;
}

- (UILabel *)resultLabel {
    if (!_resultLabel) {
        _resultLabel = [[UILabel alloc] initWithFrame:CGRectZero];
        _resultLabel.textColor = TUICallKitDynamicColor(@"callkit_recents_cell_subtitle_color", @"#888888");
        _resultLabel.font = [UIFont systemFontOfSize:12];
        _resultLabel.textAlignment =  isRTL() ? NSTextAlignmentRight : NSTextAlignmentLeft;
    }
    return _resultLabel;
}

- (UILabel *)timeLabel {
    if (!_timeLabel) {
        _timeLabel = [[UILabel alloc] initWithFrame:CGRectZero];
        _timeLabel.textColor = TUICallKitDynamicColor(@"callkit_recents_cell_time_color", @"#BBBBBB");
        _timeLabel.font = [UIFont systemFontOfSize:12];
        _timeLabel.textAlignment =  isRTL() ? NSTextAlignmentLeft : NSTextAlignmentRight;
    }
    return _timeLabel;
}

- (UIButton *)moreButton {
    if (!_moreButton) {
        _moreButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_moreButton setImage:[TUICallingCommon getBundleImageWithName:@"ic_recents_more"] forState:UIControlStateNormal];
        _moreButton.imageView.contentMode = UIViewContentModeScaleAspectFit;
    }
    return _moreButton;
}

@end

File: iOS/TUIKit/TUICallKit/TUICallKit-Swift/View/Component/Recents/TUICallRecordCallsCell.swift
//
//  TUICallRecordCallsCell.swift
//
//
//  Created by vincepzhang on 2023/8/28.
//

import Foundation
import UIKit
import TUICallEngine

class TUICallRecordCallsCell: UITableViewCell {
    
    private let faceURLObserver = Observer()
    private let titleObserver = Observer()
    
    typealias TUICallRecordCallsCellMoreBtnClickedHandler = () -> Void
    var moreBtnClickedHandler: TUICallRecordCallsCellMoreBtnClickedHandler = {}
    
    private var isViewReady = false
    private var viewModel: TUICallRecordCallsCellViewModel = TUICallRecordCallsCellViewModel(TUICallRecords())
    
    private lazy var avatarImageView: UIImageView = {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFit
        imageView.layer.cornerRadius = 20
        imageView.clipsToBounds = true
        return imageView
    }()
    
    private lazy var titleLabel: UILabel = {
        let label = UILabel()
        label.numberOfLines = 1
        label.font = UIFont(name: "PingFangHK-Semibold", size: 14)
        label.textAlignment = TUICoreDefineConvert.getIsRTL() ? .right : .left
        return label
    }()
    
    private lazy var mediaTypeImageView: UIImageView = {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFit
        return imageView
    }()
    
    private lazy var resultLabel: UILabel = {
        let label = UILabel()
        label.textColor = TUICoreDefineConvert.getTUICallKitDynamicColor(colorKey: "callkit_recents_cell_subtitle_color",
                                                                         defaultHex: "#888888")
        label.font = UIFont.systemFont(ofSize: 12)
        label.textAlignment = TUICoreDefineConvert.getIsRTL() ? .right : .left
        return label
    }()
    
    private lazy var timeLabel: UILabel = {
        let label = UILabel()
        label.textColor = TUICoreDefineConvert.getTUICallKitDynamicColor(colorKey: "callkit_recents_cell_time_color",
                                                                         defaultHex: "#BBBBBB")
        label.font = UIFont.systemFont(ofSize: 12)
        label.textAlignment = TUICoreDefineConvert.getIsRTL() ? .left : .right
        return label
    }()
    
    private lazy var moreButton: UIButton = {
        let button = UIButton(type: .custom)
        button.setImage(TUICallKitCommon.getBundleImage(name: "ic_recents_more"), for: .normal)
        button.imageView?.contentMode = .scaleAspectFit
        return button
    }()
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        contentView.backgroundColor = TUICoreDefineConvert.getTUICallKitDynamicColor(colorKey: "callkit_recents_cell_bg_color",
                                                                                     defaultHex: "#FFFFFF")
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        unregisterObserve()
    }
    
    override func didMoveToWindow() {
        super.didMoveToWindow()
        if isViewReady {
            return
        }
        isViewReady = true
        constructViewHierarchy()
        activateConstraints()
        bindInteraction()
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()
    }
    
    private func constructViewHierarchy() {
        contentView.addSubview(avatarImageView)
        contentView.addSubview(titleLabel)
        contentView.addSubview(mediaTypeImageView)
        contentView.addSubview(resultLabel)
        contentView.addSubview(timeLabel)
        contentView.addSubview(moreButton)
    }
    
    private func activateConstraints() {
        avatarImageView.snp.makeConstraints { make in
            make.centerY.equalTo(contentView)
            make.leading.equalTo(contentView).offset(16)
            make.width.height.equalTo(40)
        }
        titleLabel.snp.makeConstraints { make in
            make.top.equalTo(contentView).offset(14)
            make.leading.equalTo(avatarImageView.snp.trailing).offset(8)
            make.trailing.lessThanOrEqualTo(timeLabel.snp.leading).offset(-20)
        }
        mediaTypeImageView.snp.makeConstraints { make in
            make.bottom.equalTo(contentView).offset(-14)
            make.leading.equalTo(avatarImageView.snp.trailing).offset(8)
            make.width.equalTo(19)
            make.height.equalTo(12)
        }
        resultLabel.snp.makeConstraints { make in
            make.centerY.equalTo(mediaTypeImageView)
            make.leading.equalTo(mediaTypeImageView.snp.trailing).offset(4)
            make.width.equalTo(100)
        }
        timeLabel.snp.makeConstraints { make in
            make.centerY.equalTo(contentView)
            make.trailing.equalTo(moreButton.snp.leading).offset(-4)
            make.width.equalTo(100)
        }
        moreButton.snp.makeConstraints { make in
            make.centerY.equalTo(contentView)
            make.trailing.equalTo(-8)
            make.width.height.equalTo(24)
        }
    }
    
    private func bindInteraction() {
        moreButton.addTarget(self, action: #selector(moreButtonClick(_:)), for: .touchUpInside)
    }
    
    func configViewModel(_ viewModel: TUICallRecordCallsCellViewModel) {
        self.viewModel = viewModel
        registerObserve()
        
        titleLabel.text = viewModel.titleLabelStr.value
        avatarImageView.sd_setImage(with: URL(string: viewModel.faceURL.value), placeholderImage: viewModel.avatarImage)
        resultLabel.text = viewModel.resultLabelStr
        timeLabel.text = viewModel.timeLabelStr
        mediaTypeImageView.image = TUICallKitCommon.getBundleImage(name: viewModel.mediaTypeImageStr)
        
        if viewModel.callRecord.result == .missed {
            titleLabel.textColor = UIColor.red
        } else {
            titleLabel.textColor = TUICoreDefineConvert.getTUICallKitDynamicColor(colorKey: "callkit_recents_cell_title_color",
                                                                                  defaultHex: "#000000")
        }
    }
    
    func registerObserve() {
        viewModel.faceURL.addObserver(faceURLObserver) { [weak self] _, _ in
            guard let self = self else { return }
            self.avatarImageView.sd_setImage(with: URL(string: self.viewModel.faceURL.value),
                                             placeholderImage: self.viewModel.avatarImage)
        }
        
        viewModel.titleLabelStr.addObserver(titleObserver) { [weak self] _, _ in
            guard let self = self else { return }
            self.titleLabel.text = self.viewModel.titleLabelStr.value
        }
    }
    
    func unregisterObserve() {
        viewModel.faceURL.removeObserver(faceURLObserver)
        viewModel.titleLabelStr.removeObserver(titleObserver)
    }
    
    @objc private func moreButtonClick(_ button: UIButton) {
        moreBtnClickedHandler()
    }
}
